packages/webapp/src/components/Form/Unit/useUnit.jsx

```diff
@@ -0,0 +1,212 @@
+import { useCallback, useEffect, useState, useMemo } from 'react';
+import PropTypes from 'prop-types';
+import { get, useFormState } from 'react-hook-form';
+import {
+  area_total_area,
+  getDefaultUnit,
+  roundToTwoDecimal,
+} from '../../../util/convert-units/unit';
+import { convert } from '../../../util/convert-units/convert';
+import { getUnitOptionMap } from '../../../util/convert-units/getUnitOptionMap';
+
+const getOptions = (unitType = area_total_area, system) => {
+  return unitType[system].units.map((unit) => getUnitOptionMap()[unit]);
+};
+
+const useUnit = ({
+  disabled,
+  name,
+  displayUnitName,
+  hookFormSetValue,
+  hookFormGetValue,
+  hookFromWatch,
+  defaultValue,
+  system,
+  control,
+  unitType,
+  from: defaultValueUnit,
+  to,
+  required,
+  optional,
+  mode,
+  max,
+}) => {
+  const onClear = () => {
+    setVisibleInputValue('');
+    hookFormSetHiddenValue('', { shouldClearError: !optional, shouldValidate: optional });
+  };
+
+  const [showError, setShowError] = useState(false);
+  const [isDirty, setDirty] = useState(false);
+  const { errors } = useFormState({ control });
+  const error = get(errors, name);
+
+  useEffect(() => {
+    setShowError(!!error && !disabled && isDirty);
+  }, [error]);
+
+  const { displayUnit, displayValue, options, databaseUnit, isSelectDisabled, measure } =
+    useMemo(() => {
+      const databaseUnit = defaultValueUnit ?? unitType.databaseUnit;
+      const options = getOptions(unitType, system);
+      const hookFormValue = hookFormGetValue(name);
+      const value = hookFormValue || (hookFormValue === 0 ? 0 : defaultValue);
+      const isSelectDisabled = options.length <= 1;
+      const measure = convert().describe(databaseUnit)?.measure;
+      return to && convert().describe(to)?.system === system
+        ? {
+            displayUnit: to,
+            displayValue:
+              defaultValue && roundToTwoDecimal(convert(value).from(databaseUnit).to(to)),
+            options,
+            databaseUnit,
+            isSelectDisabled,
+            measure,
+          }
+        : {
+            ...getDefaultUnit(unitType, value, system, databaseUnit),
+            options,
+            databaseUnit,
+            isSelectDisabled,
+            measure,
+          };
+    }, []);
+
+  const hookFormUnitOption = hookFromWatch(displayUnitName);
+  const hookFormUnit = databaseUnit;
+  useEffect(() => {
+    if (typeof hookFormUnitOption === 'string' && getUnitOptionMap()[hookFormUnitOption]) {
+      hookFormSetValue(displayUnitName, getUnitOptionMap()[hookFormUnitOption]);
+    }
+  }, []);
+  useEffect(() => {
+    if (hookFormUnit && convert().describe(hookFormUnit)?.system !== system && measure !== 'time') {
+      hookFormSetValue(displayUnitName, getUnitOptionMap()[displayUnit]);
+    }
+  }, [hookFormUnit]);
+
+  useEffect(() => {
+    !hookFormGetValue(displayUnitName) &&
+      hookFormSetValue(displayUnitName, getUnitOptionMap()[displayUnit]);
+  }, []);
```

Comment:
These three useEffect are probably wrong.

Reply:
These three useEffect are probably wrong.
When the first effect hook set the initial unit option, second effect hook won't get proper hookFormUnit if block. hookFormUnit will update on next render.

Reply:
It's cleaner to write effect hooks + set values in terms of state sync conditions.
for example, 
convert(hookFormValue).from(databaseUnit).to(hookFormUnit) === displayValue should always hold.
Then there should be 3 effect hooks that listen to hookFormValue/hookFormUnit/displayValue, then setHookFormValue/displayValue so that the relationship holds.

I just don't know how to make it clean.